solvers/geomloss.py

```diff
@@ -0,0 +1,109 @@
+from benchopt import BaseSolver, safe_import_context
+from benchopt.stopping_criterion import SufficientProgressCriterion
+
+# Protect the import with `safe_import_context()`. This allows:
+# - skipping import to speed up autocompletion in CLI.
+# - getting requirements info when all dependencies are not installed.
+with safe_import_context() as import_ctx:
+    import jax.numpy as jnp
+    import numpy as np
+
+    # Import Geomloss which is based on pytorch.
+    import torch
+    from geomloss.sinkhorn_divergence import log_weights, sinkhorn_loop
+    from geomloss.sinkhorn_samples import cost_routines, softmin_tensorized
+
+    # Using OTT to get an output compatible with the objective.
+    import ott
+    from ott.geometry import pointcloud
+    from ott.problems.linear import linear_problem
+    from ott.solvers.linear.sinkhorn import SinkhornOutput
+
+
+class Solver(BaseSolver):
+
+    # Name to select the solver in the CLI and to display the results.
+    name = 'GeomLoss'
+
+    install_cmd = 'conda'
+    requirements = ['torch', 'pykeops', 'pip:geomloss', 'ott-jax']
+
+    # List of parameters for the solver. The benchmark will consider
+    # the cross product for each key in the dictionary.
+    parameters = {
+        'reg': [1e-2, 1e-1],
+        'use_gpu': [True],
+    }
+
+    stopping_criterion = SufficientProgressCriterion(patience=50)
+
+    def set_objective(self, x, a, y, b):
+        # Create a ott problem based on jax to compute the output \
+        # of the solver.
+        x_jax, y_jax, a_jax, b_jax = map(
+            lambda x: jnp.array(x), (x, y, a, b)
+        )
+        self.ot_prob = linear_problem.LinearProblem(
+            pointcloud.PointCloud(
+                x_jax, y_jax, epsilon=self.reg,
+                cost_fn=ott.geometry.costs.SqPNorm(p=2)
+            ), a_jax, b_jax,
+        )
+
+        # Store the problem in torch to use GeomLoss.
+        # Use the GPU when it is available.
+        device = (
+            'cuda' if torch.cuda.is_available() and self.use_gpu
+            else 'cpu'
+        )
```

Comment:
I would say it is better to skip the solver if `use_gpu` is `True` and no GPU support.

Reply:
sure